docker/importer/importer.py

```diff
@@ -313,11 +313,98 @@ def _vuln_ids_from_gcs_blob(self, client: storage.Client,
+    - it's new/modified since last run and the hash has changed
+    - the importer is reimporting the entire source
+      - ignore_last_import_time is True
+    - the record passes OSV JSON Schema validation
+
+    Usually an OSV file has a single vulnerability in it, but it is permissible
+    to have more than one, hence it returns a list of tuples.
+
+    This is runnable in parallel using concurrent.futures.ThreadPoolExecutor
+
+    Args:
+      storage_client: a storage.Client() to use for retrieval of the blob
+      ndb_client: an ndb.Client() to use for Data Store access
+      source_repo: the osv.SourceRepository the blob relates to
+      blob: the storage.Blob object to operate on
+
+    Raises:
+      jsonschema.exceptions.ValidationError when self._strict_validation is True
+      input fails OSV JSON Schema validation
+
+    Returns:
+      a list of one or more tuples of (hash, vulnerability) (from the
+      Vulnerability proto) or None when the blob has an unexpected name
+    """
+    if not _is_vulnerability_file(source_repo, blob.name):
+      return None
+
+    utc_last_update_date = source_repo.last_update_date.replace(
+        tzinfo=datetime.timezone.utc)
+
+    if not ignore_last_import_time and \
+        blob.updated is not None and \
+        not blob.updated > utc_last_update_date:
+      return None
+
+    # The record in GCS appears to be new/changed, examine further.
+    logging.info('Bucket entry triggered for %s/%s', source_repo.bucket,
+                 blob.name)
+
+    # Download in a blob generation agnostic way to cope with the blob
+    # changing between when it was listed and now (if the generation doesn't
+    # match, retrieval fails otherwise).
+    blob_bytes = storage.Blob(
+        blob.name, blob.bucket,
+        generation=None).download_as_bytes(storage_client)
+
+    # This is the atypical execution path (when reimporting is triggered)
+    if ignore_last_import_time:
+      blob_hash = osv.sha256_bytes(blob_bytes)
+      if self._strict_validation:
+        # This may raise a jsonschema.exceptions.ValidationError
+        _ = osv.parse_vulnerabilities_from_data(
+            blob_bytes,
+            os.path.splitext(blob.name)[1],
+            strict=self._strict_validation)
+      return blob_hash, blob.name
+
+    # This is the typical execution path (when reimporting not triggered)
+    with ndb_client.context():
+      # This may raise a jsonschema.exceptions.ValidationError
+      vulns = osv.parse_vulnerabilities_from_data(
+          blob_bytes,
+          os.path.splitext(blob.name)[1],
+          strict=self._strict_validation)
+      for vuln in vulns:
+        bug = osv.Bug.get_by_id(vuln.id)
+        # Check if the bug has been modified since last import
+        if bug is None or \
+                bug.import_last_modified != vuln.modified.ToDatetime():
+          blob_hash = osv.sha256_bytes(blob_bytes)
+          # Is this returning prematurely?
```

Comment:
I don't think so. We cant probably just make this a comment to explain this instead:

We want to tigger a re-import of the entire GCS blob if **any** vulns in the blob is:
  - Has a new ID we haven't seen before, or
  - The vuln has a different modified time.

Reply:
Thanks, I've tweaked comments and the docstring accordingly.